### PR TITLE
Remove localhost provider type

### PR DESCRIPTION
**Pull request by rekmarks**
### Summary

Recent changes to the network controller broke support for providers of type `localhost` - i.e. `Localhost 8545` - in the network dropdown.

This PR fixes our support for the `Localhost 8545` network by converting its provider type to `rpc`, and removing the `localhost` provider type completely.

If this gets merged, users will have to configure the `chainId` value of the `Localhost 8545` network if it's anything other than the Ganache default, `0x539`. We'll have to communicate that.

### Changes

- Delete `localhost` provider type and all references thereto
- Change the hardcoded Ganache network ID to 1337 (hexadecimal `0x539`)
  - This is the Ganache default. We were previously using 5777, to no conceivable benefit.
  - Since the `Localhost 8545` network is now a normal custom RPC network, the `chainId` has to be hardcoded. Most users will use the default Ganache config, and if we do the same everything works out of the box.
- Add the `Localhost 8545` network to the first time state
  - If we want to continue shipping this out of the box, this is what we have to do.
- Add migration to add `Localhost 8545` network to the preferences controller `frequentRpcListDetail`
  - Adds it to the top of the list, which will make `Localhost 8545` appear in the same place in the network dropdown
- Delete `createLocalhostClient`, add e2e test functionality to `createJsonRpcClient`
  - Basically, `createJsonRpcClient` now has some `process.env.IN_TEST` conditionals previously in `createLocalhostClient`.

### CODE REVIEW

File: app/scripts/background.js [hunk 1]
```diff
@@ -41,8 +41,7 @@ import {
 } from './lib/enums'
 /* eslint-enable import/first */
 
-// METAMASK_TEST_CONFIG is used in e2e tests to set the default network to localhost
-const firstTimeState = { ...rawFirstTimeState, ...global.METAMASK_TEST_CONFIG }
+const firstTimeState = { ...rawFirstTimeState }
```
Comment (rekmarks) [inline app/scripts/background.js:44]: I couldn't find a reference to `METAMASK_TEST_CONFIG`, so out it went.

File: test/e2e/fixtures/imported-account/state.json [hunk 2]
```diff
@@ -28,13 +28,13 @@
...
       "provider": {
-        "nickname": "",
-        "rpcUrl": "",
-        "chainId": "0x1691",
+        "nickname": "Localhost 8545",
+        "rpcUrl": "http://localhost:8545",
+        "chainId": "0x539",
         "ticker": "ETH",
-        "type": "localhost"
+        "type": "rpc"
```
Comment (rekmarks) [inline test/e2e/fixtures/imported-account/state.json:37]: I'm not sure whether the fixtures need this since it's already the default provider config in tests.

File: test/unit/app/controllers/preferences-controller-test.js [hunk 3]
```diff
@@ -522,10 +522,27 @@ describe('preferences controller', function () {
   describe('adding and removing from frequentRpcListDetail', function () {
     it('should add custom RPC url to state', function () {
       preferencesController.addToFrequentRpcList('rpc_url', '0x1')
-      preferencesController.addToFrequentRpcList('http://localhost:8545', '0x1')
```
Comment (rekmarks) [inline test/unit/app/controllers/preferences-controller-test.js:525]: `addToFrequentRpcList` used to ignore `http://localhost:8545`.

File: app/scripts/migrations/048.js [hunk 4]
```diff
@@ -34,5 +36,20 @@ function transformState (state = {}) {
     delete state.NetworkController?.provider?.rpcTarget
   }
 
+  // 3.  Add localhost network to frequentRpcListDetail.
+  if (!state.PreferencesController) {
+    state.PreferencesController = {}
+  }
```
Comment (whymarrh) [inline app/scripts/migrations/048.js:42]:
```suggestion
  state.PreferencesController ??= {}
```

File: app/scripts/migrations/048.js [hunk 5]
```diff
@@ -34,5 +36,20 @@ function transformState (state = {}) {
...
+  if (!state.PreferencesController.frequentRpcListDetail) {
+    state.PreferencesController.frequentRpcListDetail = []
+  }
```
Comment (whymarrh) [inline app/scripts/migrations/048.js:45]:
```suggestion
  state.PreferencesController.frequentRpcListDetail ??= []
```

File: app/scripts/background.js (hunk 1 above)
Comment (whymarrh) [reply app/scripts/background.js:44]: Yeah I think this might have been a remnant of an old e2e setup

File: app/scripts/controllers/network/createJsonRpcClient.js [hunk 6]
```diff
@@ -7,19 +8,33 @@ import createBlockTrackerInspectorMiddleware from 'eth-json-rpc-middleware/block
...
   const blockProvider = providerFromMiddleware(fetchMiddleware)
-  const blockTracker = new BlockTracker({ provider: blockProvider })
+  const blockTracker = new BlockTracker({
+    ...blockTrackerOpts,
+    provider: blockProvider,
+  })
 
-  const networkMiddleware = mergeMiddleware([
+  const middlewares = [
     createChainIdMiddleware(chainId),
```
Comment (whymarrh) [inline app/scripts/controllers/network/createJsonRpcClient.js:25]: It might be simpler to define a `testMiddlewares` array and spread that onto this list:

```suggestion
    ...testMiddlewares,
    createChainIdMiddleware(chainId),
```

This would simplify the diff but also keep all the test related code in one place—we could define `testMiddlewares` up alongside `inTest` and `blockTrackerOpts`.

**Comment by danjm** [comment]
> users will have to configure the chainId value of the Localhost 8545 network if it's anything other than the Ganache default, 0x539

should this go in documentation somewhere?

**Comment by danjm** [comment]
code looks good, I will QA now

File: app/scripts/migrations/048.js (hunk 4 above)
Comment (rekmarks) [reply app/scripts/migrations/048.js:42]: For posterity: this wasn't done because we don't have nullish assignment yet.

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 6 above)
Comment (rekmarks) [reply app/scripts/controllers/network/createJsonRpcClient.js:31]: Done in: a2caf2e

File: app/scripts/migrations/048.js (hunk 5 above)
Comment (rekmarks) [reply app/scripts/migrations/048.js:45]: For posterity: this wasn't done because we don't have nullish assignment yet.

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 6 above)
Comment (whymarrh) [reply app/scripts/controllers/network/createJsonRpcClient.js:31]: We can restore the single call to `mergeMiddleware` as well, that doesn't need to be split out anymore

Comment (rekmarks) [reply app/scripts/controllers/network/createJsonRpcClient.js:31]: Fixed in: a665c2f

**Comment by danjm** [comment]
My QA looked good. Erik is QAing the migration

File: app/scripts/controllers/network/createJsonRpcClient.js [hunk 7]
```diff
@@ -7,19 +8,34 @@ import createBlockTrackerInspectorMiddleware from 'eth-json-rpc-middleware/block
 import providerFromMiddleware from 'eth-json-rpc-middleware/providerFromMiddleware'
 import BlockTracker from 'eth-block-tracker'
 
+const inTest = process.env.IN_TEST === 'true'
+const blockTrackerOpts = inTest
+  ? { pollingInterval: 1000 }
```
Comment (danjm) [inline app/scripts/controllers/network/createJsonRpcClient.js:13]: Before this PR, these opts were not just included when `inTest` https://github.com/MetaMask/metamask-extension/pull/9551/files#diff-5d7a765d6c410d375ee6359517ee8333L14

Why is that change happening now?

Comment (Gudahtt) [reply app/scripts/controllers/network/createJsonRpcClient.js:13]: Previously we were using the old localhost provider in test, which did have this option set. See below, the deleted file `createLocalHostClient.js`.

The only change here is setting the polling time to 1s instead of the default of 20s when in test.

Comment (rekmarks) [reply app/scripts/controllers/network/createJsonRpcClient.js:13]: The link doesn't take me to a specific line, but I assume you're referring to the content of `createLocalhostClient`: https://github.com/MetaMask/metamask-extension/blob/6409caa0814d228e42b005992b96b55b92948431/app/scripts/controllers/network/createLocalhostClient.js

To answer your question, `createLocalhostClient` was basically `createJsonRpcClient`, with some optimizations/configuration options for `localhost` networks. We don't _need_ to make those optimizations, so now we ignore them in production.

However, our e2e tests rely on those optimizations. The [default polling interval is 20 seconds](https://github.com/MetaMask/eth-block-tracker/blob/master/src/polling.js#L14), which is longer than many of our e2e test timeouts.

I haven't confirmed whether the estimate gas delay or polling interval alone are sufficient to get e2e tests reliably passing, but e2e tests do fail in the absence of both. Given how flaky our e2e tests are, the safest bet is probably to include both options.

**Comment by danjm** [comment]
A UX change with this PR is that the 'Localhost 8545' can be deleted and edited, whereas it couldn't previously. I think this is okay? Do we want to prevent deletion?

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 7 above)
Comment (rekmarks) [reply app/scripts/controllers/network/createJsonRpcClient.js:13]: Now, this will degrade development experience somewhat on Ganache networks (since the polling interval is the same as on a "real" network), but we can address that after 8.1.

File: app/scripts/migrations/048.js [hunk 8]
```diff
@@ -24,7 +25,8 @@ function transformState (state = {}) {
   delete state.NetworkController?.settings
 
   // 2. Delete NetworkController.provider or rename rpcTarget key
-  if (state.NetworkController?.provider?.type === 'rpc') {
+  const providerType = state.NetworkController?.provider?.type
+  if (providerType === 'rpc' || providerType === 'localhost') {
```
Comment (danfinlay) [inline app/scripts/migrations/048.js:29]: Will this bump users of a custom RPC onto main-net? If so, I could imagine users being surprised that what they thought was test-ether or something else is now main-net ether. We do show visual indications of the current network, so this is "theoretically safe", but I could imagine problems.

File: test/unit/migrations/048-test.js [hunk 9]
```diff
@@ -60,14 +76,40 @@ describe('migration #48', function () {
...
     assert.deepEqual(newStorage.data, {
+      ...expectedPreferencesState,
+      NetworkController: {
+        foo: 'bar',
+      },
+      foo: 'bar',
+    })
+  })
+
+  it('should delete NetworkController.provider if the type is "localhost"', async function () {
```
Comment (danfinlay) [inline test/unit/migrations/048-test.js:87]: I guess the alternative would be to migrate localhost types, but my guess is we don't have a chainId recorded for these so we can't easily? Could we treat them the same way we do other RPCs with no chainId?

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 7 above)
Comment (danfinlay) [reply app/scripts/controllers/network/createJsonRpcClient.js:13]: Keeping polling rate more dynamic is already a good thing for us to aim for, as different EVM blockchains have different block frequencies.

[Opened an issue](https://github.com/MetaMask/metamask-extension/issues/9556)

File: app/scripts/controllers/network/createJsonRpcClient.js [hunk 10]
```diff
@@ -32,3 +47,16 @@ function createChainIdMiddleware (chainId) {
...
 }
+
+/**
+ * For use in tests only.
+ * Adds a delay to `eth_estimateGas` calls.
+ */
+function createEstimateGasDelayMiddleware () {
+  return createAsyncMiddleware(async (req, _, next) => {
+    if (req.method === 'eth_estimateGas') {
+      await new Promise((resolve) => setTimeout(resolve, 2000))
```
Comment (danfinlay) [inline app/scripts/controllers/network/createJsonRpcClient.js:59]: This `resolve` call doesn't need to return a value?

**Review by danfinlay — COMMENTED** [review]
This mostly all looks good, I had a couple small comment/questions.

**Comment by rekmarks** [comment]
@danjm I definitely think it's okay, but worth floating to a wider audience.

**Comment by Gudahtt** [comment]
This seems like it could be a generally good change - removes a lot of localhost-specific code, which is good. But what problem is it fixing, and why do we want to do this now? It seems like a pretty substantial change to make at this time.

I understand that the localhost provider is "broken", but how is it broken? Is a simpler way of fixing it perhaps, just for now?

File: app/scripts/controllers/network/createJsonRpcClient.js (hunk 10 above)
Comment (rekmarks) [reply app/scripts/controllers/network/createJsonRpcClient.js:59]: Nah, as written it's just a `Promise<undefined>`, and we of course don't do anything with the return value anyway.

File: test/unit/migrations/048-test.js (hunk 9 above)
Comment (rekmarks) [reply test/unit/migrations/048-test.js:87]: AFAIK, the only network with a `localhost` provider type is the `Localhost 8545` network we hard-coded into the extension.

Comment (rekmarks) [reply test/unit/migrations/048-test.js:87]: So, since we know everything about that network, we're effectively migrating the sole instance of `localhost` networks.

File: app/scripts/migrations/048.js (hunk 8 above)
Comment (rekmarks) [reply app/scripts/migrations/048.js:29]: Yeah, it does. We could switch the user to a test network instead of Mainnet in the migration, or add another alert.

**Comment by rekmarks** [comment]
@Gudahtt the `localhost` provider is completely broken on `develop`, as in it can't be used at all. The alternative would be #9550, although that has other drawbacks.

File: app/scripts/first-time-state.js [hunk 11]
```diff
@@ -10,6 +10,17 @@
  */
 const initialState = {
   config: {},
+  PreferencesController: {
```
Comment (Gudahtt) [inline app/scripts/first-time-state.js:13]: Ni: It would seem safer to add this in the migration, rather than here. Anything added here will be subject to all of the migrations for new installations.

If we've tested this specific addition and it's OK, then maybe this is non-blocking, but this isn't a pattern we should follow often I don't think.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by kumavis — COMMENTED** [review]
somehow I didn't submit this review in time

File: test/unit/migrations/048-test.js [hunk 12]
```diff
@@ -111,16 +154,39 @@ describe('migration #48', function () {
     }
 
     const newStorage = await migration48.migrate(oldStorage)
-    assert.deepEqual(oldStorage.data, newStorage.data)
+    assert.deepEqual(
+      { ...oldStorage.data, ...expectedPreferencesState },
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply test/unit/migrations/048-test.js:158]: Hah, I waffled on this and made the wrong call. Looking at the various values, we can tell that it's fine in this case, but I nevertheless made the change here: https://github.com/MetaMask/metamask-extension/pull/9567
